### PR TITLE
[hotfix] Adjust to Herokus subdomain change 

### PR DESCRIPTION
**Pull request by Deradon**
## Issue

Heroku recently introduced [subdomain changes](https://devcenter.heroku.com/changelog-items/2640).

> _Starting June 14, 2023, new apps get a randomly generated 12-character identifier appended to the subdomain portion of the built-in APPNAME.herokuapp.com domains._

That's a problem for reina as it's guessing/generating URLs.

## Solution

Getting the URL from the Heroku API.

## Challenges

We're setting up ENV vars for our services. E.g. `frontend` heroku app might have an ENV var `API_URL`.
Previously we were guessing the value and were able to set this ENV var w/o having the `backend` already created.
As we can not guess the value anymore, the `backend` service has to exist to get the correct value.

So we had to change how reina works:

* First: Create all heroku apps
* Second: Deploy and set ENV vars (as all heroku apps exist already)

## Out of scope

For this PR anything beyond hotfixing is considered out of scope. reina could benefit a lot from further refactoring.

### CODE REVIEW

File: lib/reina/app.rb [hunk 1]
```diff
@@ -36,10 +44,8 @@ def create_app
         'stack' => project.fetch(:stack, DEFAULT_STACK)
       }
 
-      if team = project[:team]
-        heroku.team_app.create(
-          params.merge('team' => team)
-        )
+      @info = if (team = project[:team])
```
Comment (Deradon) [inline lib/reina/app.rb:47]: No we store the return value from the API, which includes the correct URL. (see [docs](https://github.com/heroku/platform-api#usage-guide))

File: lib/reina/app.rb [hunk 2]
```diff
@@ -68,32 +74,37 @@ def add_buildpacks
...
             next
           end
 
-          source, var = h[:from].split('#')
+          source, var = from.split('#')
           source_app_name = app_name_for(source)
 
           if var == 'url'.freeze
-            config_vars[h[:to]] = "https://#{domain_name_for(source_app_name)}"
+            info = heroku.app.info(source_app_name)
```
Comment (Deradon) [inline lib/reina/app.rb:103]: :warning: We fetch the app-info (and therefore the correct `web_url`) during runtime. (this request the heroku-app to exist ofc)

FYI: Config looks something like

```json
{
  "frontend": {
    "github": "frontend",
    "pipeline": "frontend",
    "team": "honeypot",
    "stack": "heroku-22",
    "default_branch": "main",
    "config_vars": {
      "from": "staging-frontend",
      "copy": [
        {
          "from": "backend#url",
          "to": "API_URL"
        }
      ]
    }
  }
}
```

File: lib/reina/controller.rb [hunk 3]
```diff
@@ -30,34 +30,24 @@ def create_netrc
...
-    def deploy_parallel_apps!
-      Parallel.each(apps.select(&:parallel?)) do |app|
-        begin
-          deploy!(app)
-        rescue Exception => e
-          wrapped = DeploymentError.new(app, e)
-          raise_errors ? raise(wrapped) : puts(wrapped.message)
-        end
-      end
+    def create_apps!
```
Comment (Deradon) [inline lib/reina/controller.rb:41]: :warning: Creating and deploying apps split. (as the `app.set_env_vars` requires all apps to be created already)

File: lib/reina/github_controller.rb [hunk 4]
```diff
@@ -116,14 +104,26 @@ def deploy!(strict = false)
 
         reina.create_netrc if reina.heroku?
         reina.delete_existing_apps!
-        reina.deploy_parallel_apps!
-        reina.deploy_non_parallel_apps!
+        reina.create_apps!
+        reina.deploy_apps!
```
Comment (Deradon) [inline lib/reina/github_controller.rb:108]: :warning: As mentioned before, creating and deploying apps is done in two separate steps now.

File: lib/reina/app.rb [hunk 5]
```diff
@@ -4,29 +4,37 @@ class App
...
+    # @param heroku [PlatformAPI::Client] the Heroku API client
+    # @param name [Symbol]
+    # @param project [Hash]
+    # @param issue_number [Integer]
+    # @param branch [String]
+    #
+    # @example
+    #   heroku = PlatformAPI.connect_oauth(CONFIG[:platform_api])
+    #
+    #   app = Reina.app.new(
```
Comment (Deradon) [inline lib/reina/app.rb:18]: `Reina::App.new`

File: lib/reina/app.rb [hunk 6]
```diff
@@ -210,6 +230,7 @@ def remote_url
       "https://git.heroku.com/#{app_name}.git"
     end
 
+
```
Comment (heliosam) [inline lib/reina/app.rb:233]: 👀

**Pull request merged**